safehttp/plugins/xsrf/xsrf.go

```diff
@@ -23,141 +23,249 @@ import (
-//
-// In case of state changing requests (all except GET, HEAD and OPTIONS), the
-// interceptor checks for the presence of the XSRF token in the request body
-// (expected to have been injected) and validates it.
 func (it *Interceptor) Before(w *safehttp.ResponseWriter, r *safehttp.IncomingRequest, _ safehttp.InterceptorConfig) safehttp.Result {
-	needsValidation := !statePreservingMethods[r.Method()]
-	cookieID, err := r.Cookie(cookieIDKey)
+	code := it.c.Check(r)
+	if code != safehttp.StatusOK {
+		return w.WriteError(code)
+	}
+	return safehttp.NotWritten()
+}
+
+func (it *Interceptor) Commit(w *safehttp.ResponseWriter, r *safehttp.IncomingRequest, resp safehttp.Response, _ safehttp.InterceptorConfig) safehttp.Result {
+	data, err := it.g.Generate(r)
 	if err != nil {
-		if needsValidation {
-			return w.WriteError(safehttp.StatusForbidden)
-		}
-		cookieID, err = addCookieID(w)
-		if err != nil {
-			// An error is returned when the plugin fails to Set the Set-Cookie
-			// header in the response writer as this is a server misconfiguration.
-			return w.WriteError(safehttp.StatusInternalServerError)
-		}
+		return w.WriteError(safehttp.StatusInternalServerError)
 	}
+	err = it.i.Inject(resp, w, data)
+	if err != nil {
+		return w.WriteError(safehttp.StatusInternalServerError)
+	}
+	return safehttp.Result{}
+}
 
-	actionID := r.URL.Path()
-	if needsValidation {
-		f, err := r.PostForm()
-		if err != nil {
-			// We fallback to checking whether the form is multipart. Both types
-			// are valid in an incoming request as long as the XSRF token is
-			// present.
-			mf, err := r.MultipartForm(32 << 20)
-			if err != nil {
-				return w.WriteError(safehttp.StatusBadRequest)
-			}
-			f = &mf.Form
-		}
+type Checker interface {
+	Check(r *safehttp.IncomingRequest) safehttp.StatusCode
+}
 
-		tok := f.String(TokenKey, "")
-		if f.Err() != nil || tok == "" {
-			return w.WriteError(safehttp.StatusUnauthorized)
-		}
+type Generator interface {
+	Generate(r *safehttp.IncomingRequest) (GeneratedData, error)
+}
+
+type Injector interface {
+	Inject(resp safehttp.Response, w *safehttp.ResponseWriter, data GeneratedData) error
+}
+
+type GeneratedData interface{}
+
+type defaultGenerator struct {
+	secretAppKey string
+	cookieIDKey  string
+}
+type defaultData struct {
+	cookieID  *safehttp.Cookie
+	token     string
+	setCookie bool
+}
 
-		if ok := xsrftoken.Valid(tok, it.SecretAppKey, cookieID.Value(), actionID); !ok {
-			return w.WriteError(safehttp.StatusForbidden)
+func (g defaultGenerator) Generate(r *safehttp.IncomingRequest) (GeneratedData, error) {
+	data := defaultData{}
+	c, err := r.Cookie(g.cookieIDKey)
+	if err != nil {
+		buf := make([]byte, 20)
+		if _, err := rand.Read(buf); err != nil {
+			return nil, fmt.Errorf("crypto/rand.Read: %v", err)
 		}
+		c = safehttp.NewCookie(g.cookieIDKey, base64.StdEncoding.EncodeToString(buf))
+		c.SetSameSite(safehttp.SameSiteStrictMode)
+		data.setCookie = true
 	}
+	data.cookieID = c
```

Comment:
Creating cookies should be done in `Injector`, not `Generator`.

For instance, how weird would it be to set cookie strictness in `Generator`?